### PR TITLE
[DNM]MVCC GC

### PR DESCRIPTION
**Pull request by c4pt0r**

### CODE REVIEW

**Comment by c4pt0r** [comment]
@ngaut PTAL

**Comment by c4pt0r** [comment]
I'm adding tests.

File: kv/gc.go [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package kv
+
+type GC interface {
```
Comment (ngaut) [inline kv/gc.go]: Is Compactor a better name?

File: store/localstore/local_gc.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+	gcInterval        = 3 * time.Second
+)
+
+type localstoreGC struct {
+	mu         sync.Mutex
+	recentKeys map[string]struct{}
+	stopChan   chan struct{}
+	delChan    chan kv.EncodedKey
+	ticker     *time.Ticker
+	store      *dbStore
```
Comment (c4pt0r) [inline store/localstore/local_gc.go]: use engine.DB instead of dbStore

File: store/localstore/local_gc.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+	}()
+
+	// delete worker
+	go func() {
+		cnt := 0
+		batch := gc.store.db.NewBatch()
+		for k := range gc.delChan {
+			cnt++
+			batch.Delete(k)
+			if cnt == 100 {
```
Comment (c4pt0r) [inline store/localstore/local_gc.go]: use gcBatchDeleteSize

File: store/localstore/local_gc.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+				cnt = 0
+			}
+		}
+	}()
+}
+
+func (gc *localstoreGC) Stop() {
+	gc.stopChan <- struct{}{}
+	gc.ticker.Stop()
+	close(gc.stopChan)
```
Comment (c4pt0r) [inline store/localstore/local_gc.go]: close delChan

File: store/localstore/local_gc.go [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+	var ret []kv.EncodedKey
+	for it.Next() {
+		if kv.EncodedKey(it.Key()).Cmp(endKey) < 0 {
+			ret = append(ret, kv.EncodedKey(it.Key()))
+		}
+	}
+	return ret, nil
+}
+
+func (gc *localstoreGC) Do(k kv.Key) {
```
Comment (ngaut) [inline store/localstore/local_gc.go]: d/Do/Collect/g

File: store/localstore/local_gc.go [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+			batch.Delete(k)
+			if cnt == 100 {
+				// batch delete
+				err := gc.store.db.Commit(batch)
+				if err != nil {
+					// Ignore error
+					log.Error(err)
+				}
+				log.Debug("batch gc done")
+				batch = gc.store.newBatch()
```
Comment (ngaut) [inline store/localstore/local_gc.go]: can we reuse batch ?

File: kv/gc.go [hunk 7]
```diff
@@ -0,0 +1,7 @@
+package kv
```
Comment (qiuyesuifeng) [inline kv/gc.go]: Add license header.

File: kv/gc.go [hunk 8]
```diff
@@ -0,0 +1,7 @@
...
+
+type GC interface {
+	OnGet(k Key)
+	OnSet(k Key)
```
Comment (ngaut) [inline kv/gc.go]: missing OnDelete

File: store/localstore/local_gc.go [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+				// Do GC
+				for k, _ := range m {
+					gc.Do([]byte(k))
+				}
+			}
+		}
+	}()
+
+	// delete worker
+	go func() {
```
Comment (ngaut) [inline store/localstore/local_gc.go]: Please use named method instead of closure.

File: store/localstore/local_gc.go [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+	cnt        int
+}
+
+func (gc *localstoreGC) OnSet(k kv.Key) {
+	gc.mu.Lock()
+	defer gc.mu.Unlock()
+	gc.recentKeys[string(k)] = struct{}{}
+}
+
+func (gc *localstoreGC) OnGet(k kv.Key) {
```
Comment (qiuyesuifeng) [inline store/localstore/local_gc.go]: OnGet and OnSet do the same thing?

File: store/localstore/kv.go [hunk 11]
```diff
@@ -139,6 +143,7 @@ func (s *dbStore) Begin() (kv.Transaction, error) {
 func (s *dbStore) Close() error {
 	mc.mu.Lock()
 	defer mc.mu.Unlock()
+	s.gc.Stop()
```
Comment (qiuyesuifeng) [inline store/localstore/kv.go]: Call gc.Stop(), delete worker maybe not exist before u call s.db.Close(), then gc.store.db.Commit seems not work well.

**Pull request closed without merging**